### PR TITLE
Pimped-up lane hover !

### LINKED ISSUE #543: Lane connector: Hovering over lane points is difficult when cars and roads are passing over.
when connecting lanes, it can be frustrating to hover over lane markers if there is a car over it (or if there is a road over it) because the place the mouse pointer has to be to select the marker is different than the place where the marker has been rendered in a 3D setting:
Cities_VEXJKsBNIG

To fix this we need to either change the hovering functionality or the rendering functionality to match one another in a 3D setting.

### LINKED ISSUE #625: lane connector should raycast the whole lane
see #543 

in https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/543#issuecomment-549069179 @krzychu124 fixing the rendering problem.

EDIT: the current ray-casting would work just fine. 
meanwhile this issue is for making it easier to hover over lane node markers by making the whole lane hover-able.

~~This issue is for fixing the ray-casting problem.
I intend to use `GetClosestLane()`. should be a piece of cake! ~~

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #625

See https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/625#issuecomment-580706371 for pictures

- As agreed in https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/543#issuecomment-549069179 I make the whole lane hover-able. 

- Visual feedback: ~~If mouse pointer is on circle of the node lane marker then I don't highlight the whole lane. If the mouse pointer is on the lane but not on the circle, then I highlight the whole lane. This conditional display lane overlay is an answer to https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/625#issuecomment-580759773~~
selected and hovered lanes (source or target) get Sheath overlay. I didn't change the color ... just the shape.

- I revived the lane hovering code from Traffic ++

- I made performance optimization by caching mouse pointer and lane bounds. They are recalculated only if user moves the mouse ray.

fixes #543
- I fixed the vertical mismatch by intersecting the mouse ray at the place where it hits the road. 

- hitPos error: raycasting returns hitpos with error. it is as if the mouse ray penetrates the road segment a little bit. This error is partially mitigated by calling GetClosestLanePosition(). the returned vertical position still has less error and it does not cause any practical problem.

- the fixed position takes 100us to calculate. So I cached it and it only recalculated if user moves the mouse ray.

EDIT: In the old code it is possible to turn the camera so extreme that two markers are hovered. I added a check to avoid that.

EDIT: fixed lane curves:
 - TMPE does not draw lane curves for metros. I fixed that.
 - also draw lane curve when mouse is not hovering over anything.

EDIT: for monorail I no longer hover rail projections on the ground, but rather I hover the road lane it self.

EDIT: train markers are both source and target and used to be rendered with white color even if they were the source of the selection. I fixed that to use original color. see https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-583723848

EDIT: sometimes lane curves get berried under other lane curves. To mitigate this problem I render curves from selected lane on top of all others.
https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-583726108

### CODE REVIEW

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 1]
```diff
@@ -54,19 +55,94 @@ private class NodeLaneMarker {
             internal ushort SegmentId;
             internal ushort NodeId;
             internal bool StartNode;
-            internal Vector3 Position;
-            internal Vector3 SecondaryPosition;
+            internal Vector3 Position; /// projected on train
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: * terrain

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 2]
```diff
@@ -223,27 +295,29 @@ bool drawMarker
...
-                    } else {
-                        markerIsHovered = false;
-                    }
-
-                    if (markerIsHovered) {
-                        hoveredMarker = laneMarker;
-                    }
-
-                    var circleColor = laneMarker.IsTarget ? Color.white : laneMarker.Color;
+                        float hitH = TrafficManagerTool.GetAccurateHitHeight();
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: This part of the code was a mess. so I cleaned it up a bit. the logic is the same.
The main differences are:
- I intersect mouse ray with segment lane marker.
- in the old code `IsLaneMarkerHovered()` was calculated more times than necessary. Now I only call it if I know I need it.

**Comment by aubergine10** [comment]
Regarding caching: If I'm moving the mouse over the screen, it doesn't really need to recalculate on every frame. Maybe add a de-spam of 250ms between refreshes? (inperceptible to most humans)

Also, what happens if user changes camera view via keyboard (eg. cursor keys; no mouse movement)? That should invaldiate cache.

> The mouse/camera delta & de-spam caching mechanism would be super-useful for lots of other things too, for example caching of visible nodes/segments for persistent overlays, etc.

**Comment by aubergine10** [comment]
> Mouse ray will change ... I think. To be honest I don't know what exactly is mouse ray. I am checking for mouse position on terrain and THAT will change if you move the screen.

Isn't that expensive though?

There is likely a "mouse move" event on camera or somehwere else that will give a delta and that combined with check on camera position (with some 'ignore move of only few pixels') might prove a lot cheaper as a way of short-circuiting out of doing more expensive checks relating to rays.

```
// pseudo code
if (mouse_move_delta > some_amount || camera_position_change_delta > some_amount) {
   // it's worth updating
   // ideally have a de-spam timer too, to reduce refreshes if user doing lots of movement/scrolling
}
```

I think @kvakvs and @krzychu124 have looked in to some of this stuff while investigating ways to cache stuff such as speed limit overlays.

**Comment by kianzarrin** [comment]
If there is a train over passing a road the road gets priority for raycasting (this is how the code as always been). 
Cities_gFJie4fC6p

This means that if there is a train over passing  a node my code will calculate vertical hitpos at road level on the obstructing train level.
Cities_0jsfhoySRZ

Its still an improvement over old hovering algorithm because now the user can put the mouse pointer anywhere on the lane and select the marker :)

**Comment by kianzarrin** [comment]
> @aubergine10 There is likely a "mouse move" event on camera or somehwere else that will give a delta and that combined with check on camera position (with some 'ignore move of only few pixels') might prove a lot cheaper as a way of short-circuiting out of doing more expensive checks relating to rays.

Good idea. How much is considered to be tiny though? maybe we should call the OnRenderOverlay every 250ms? We need to create a issue and have a discussion on this. This is a bit too complicated to be as part of this pull request. 

> I think @kvakvs and @krzychu124 have looked in to some of this stuff while investigating ways to cache stuff such as speed limit overlays.

is there an issue number?

**Comment by kvakvs** [comment]
@kianzarrin https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/521
This has nothing to do with rays, we cached visible nodes and segments to optimize visible overlay rendering.

If you look at screen as a slice of Camera frustum, then the ray would come from Camera focus point far behind the camera, pass through the current mouse position on screen (on that slice of the camera frustum which we call screen) and hit something in the world. You can define what you want to hit using hit flags (i.e. cars, roads, terrain etc).

**Comment by aubergine10** [comment]
Lane highlight flickers when the mouse moves over the circle - can that be stopped?

> Note: Heavily compressed gif; ignore the weird mouse trail.
> ezgif com-optimize(1)

The flicker also occurs when I have a lane selected and am dragging the connector line on to an outgoing lane - it's lane highlight flickers when the connector moves over circle.

Put another way; the lane highlight should always be there while I'm over the lane, even if mouse is over the circle.

----

Lane highlight is different width to circle, and extends beyond the end of the circle which looks messy.

image

Could we make it same width as circle, and remove the tip (indicated by pink arrow) of lane highlight? That way it would be a bit neater - would look like a lane highlight with a circle at the end.

Can we make lane highlights same colour as their associated circle? Outgoing lanes already do this: white lane highlight and white circle; so it's just the incoming lanes that would need to change color.

----

When connector line is dragged away from any road, it disappears, which I found slightly disorientating.

ezgif com-optimize(1)

Is it possible to keep the connector line visible even if it's not over a segment?

**Comment by kianzarrin** [comment]
@aubergine10  regarding https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-581701191
I can cut the bezier short. the downside is that when the mouse pointer is on the node passed the 'node marker circle', user cannot select lanes. only when marker is on the 'node marker circle' or before 'node marker circle' on the segment itself, user can select lanes.

**Comment by kianzarrin** [comment]
drawing lane curve while mouse is not on segment opened a whole can of worms (not your usual kind of worms ... these ones are welcome!)
Cities_SN5VfGiWNn

lane curves were not being rendered underground at all. 
Screenshot (402)

I fixed that. I had to find curve projection on the ground:
Cities_nMOuJSwB70

My code changes the behavior of the lane markers on monorail. Its no longer possible to select a monorail lane marker by its projection on the ground:
Cities_OxyZwSUC4D

on the plus side you can chose road markers instead:
Cities_5Q8KMIm3bk

Cities_XoUYMTiByt

I ran some tests with roads, overpassing roads and trains too:
cMdkJRwjp7
Ond6mIxC8v
Cities_IN1ac6APRu

**Comment by kianzarrin** [comment]
Oh and by the way I changed the color of the lane hover and embolden them. I trimmed the ends to fix the flickering issue @aubergine10  mentioned:
0RbZf04NTe

monorail node markers seem to hide the road node markers. I created #654.

**Comment by aubergine10** [comment]
This is looking much better; still testing but will report back if I find anything.

Regarding disambiguation between road lanes and track lanes, see https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/654#issuecomment-582506554 - I think we should probably tackle that stuff separately as it's a meta issue affecting multiple tools (although any improvements in this PR still welcome).

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 3]
```diff
@@ -54,19 +57,148 @@ private class NodeLaneMarker {
...
+                    Position.y - 100f, // through all the geometry -100..100
+                    Position.y + 100f,
+                    false,
+                    false);
+            }
+
+        }
+
+        // code revived from the old Traffic++ mod : https://github.com/joaofarias/csl-traffic/blob/a4c5609e030c5bde91811796b9836aad60ddde20/CSL-Traffic/Tools/RoadCustomizerTool.cs
+        private class SegmentLaneMarker {
```
Comment (aubergine10) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: We should ideally split this stuff out in to separate UI class; see #630

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 4]
```diff
@@ -54,19 +57,148 @@ private class NodeLaneMarker {
...
+                    // if marker is projected on road plane above then modify its height
+                    pos.y = hitH;
+                }
+                Bounds bounds = new Bounds(Vector3.zero, Vector3.one * Radius) {
+                    center = pos,
+                };
+                return bounds.IntersectRay(ray);
+            }
+
+            internal void RenderOverlay(RenderManager.CameraInfo cameraInfo, Color color, bool enlarge = false) {
```
Comment (aubergine10) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: We should ideally split this stuff out in to separate UI class; see #630

**Comment by aubergine10** [comment]
> @aubergine10 regarding [#635 (comment)](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-581701191)
> I can cut the bezier short. the downside is that when the mouse pointer is on the node passed the 'node marker circle', user cannot select lanes. only when marker is on the 'node marker circle' or before 'node marker circle' on the segment itself, user can select lanes.

So there's no way to show lane outline while the circle is hovered?

----

Also, I'm getting different result to your gif in https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-582485196

image

**Comment by kvakvs** [comment]
Lifehack: Do not render a circle when an entire lane is rendered over it.

**Comment by aubergine10** [comment]
> Lifehack: Do not render a circle when an entire lane is rendered over it.

IMO circle should be rendered, just over the lane highlight.

BEHOLD! My god-tier mockup skillz:

image

1. When lane not hovered, only circle shown.
2. User hovers lane: lane highlight + circle (looks like a wire sheath) 
    > image
3. User selects lane: now wire is inserted in source sheath; target lanes indicarted by circles
4. User hovers target lane: it goes in to sheath mode (lane highlight + circle)
5. Not shown on god-tier mockup: User clicks destination = wire inserted in to that sheath

So we no longer care about whether user is hovering circle or lane, it's always based on hovering lane. If lane not hovered, then circle shown if applicable (ie. indicate it can be selected). Once lane hovered or selected, the sheath (lane outline) also appears _in addition_ to the circle.

The 'sheath' effect could be improved by making lane highlight border same color/opacity as the semi-transparent fill.

**Comment by kianzarrin** [comment]
> @aubergine10 So we no longer care about whether user is hovering circle or lane, it's always based on hovering lane. If lane not hovered, then circle shown if applicable (ie. indicate it can be selected). Once lane hovered or selected, the sheath (lane outline) also appears in addition to the circle.

I would have agreed with you if there was only one lane. but if there are multiple connected lanes your sheaths would look like Xmass and totally confusing.

I am trying to avoid highlighting the lane as an answer to https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/625#issuecomment-580759773 . The only reason I highlight the lane is because I want the user to have some feedback to know why the lane has been  hovered. This feeedback of course is not necessary if the mouse pointer is right on the marker.

I am going to slightly lengthen the lane overlay though.

PS: I use highlight and overlay interchangeably.

**Comment by kianzarrin** [comment]
How is this for sheath :) 
Cities_drnfKTikS6

**Comment by aubergine10** [comment]
That looks great!

**Comment by aubergine10** [comment]
> I would have agreed with you if there was only one lane. but if there are multiple connected lanes your sheaths would look like Xmass and totally confusing.

Only active selection (incoming lane and its connected outgoing lanes) would have sheaths.

If user chooses a different incoming lane, all the previous sheaths disappear, as we're now working with a differnt lane. So the sheaths show what's actively being worked with, differentiating it from stuff that is not actively being worked with. The sheaths should disappear whenever possible to reduce clutter.

**Comment by kianzarrin** [comment]
> @aubergine10  Only active selection (incoming lane and its connected outgoing lanes) would have sheaths.
> If user chooses a different incoming lane, all the previous sheaths disappear, as we're now working with a differnt lane. So the sheaths show what's actively being worked with, differentiating it from stuff that is not actively being worked with. The sheaths should disappear whenever possible to reduce clutter.

@kvakvs  Do you agree?

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 5]
```diff
@@ -117,7 +117,8 @@ private class NodeLaneMarker {
         private class SegmentLaneMarker {
             internal uint laneID;
             internal int laneIndex;
-            internal Bezier3 bezier;
+            internal Bezier3 raycast_bezier;
```
Comment (kvakvs) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: Naming style camelCase: `raycastBezier`

**Comment by kvakvs** [comment]
Only highlight relevant lanes at this moment in time when editing/mouse hovering.
Only show lane connector curves while the tool is active.

Looking good, though in the picture that dark red is hard to see without a contrast color outline, i'm not sure whether it will be good with outlines, maybe don't do them here, too much visual overload. Keep stuff as it is now.

**Comment by kianzarrin** [comment]
> @kvakvs  in the picture that dark red is hard to see without a contrast color outline, i'm not sure whether it will be good with outlines, maybe don't do them here, too much visual overload. Keep stuff as it is now.

I don't understand which picture? which outline? what am I suppose to do or not do? 

Are you talking about this picture? because the overlay seems pretty visible to me:
Cities_drnfKTikS6

So should I go with @aubergine10 approach?

**Comment by kvakvs** [comment]
I am pointing at dark red overlay which is good and clean. But in its brightness that color is too close to the road surface brightness which makes it dim and harder to see. If i want to create my own strong opinion i should really playtest this with those kind of crossings.

The safe approach would be to render contrast outlines like I did to lane connector curves, but i am afraid that it will be too much visual overload on screen. Maybe not.

**Comment by kianzarrin** [comment]
I implemented Aubergine approach ( it was only a couple of line of change). 
Also lane curve snaps to hovered lane as you can see in the picture bellow (if nothing is hovered then it follows mouse):
Cities_Vhqa00QTA4

Now you guys need to tell me in a clear language what exactly do you want? Implementation wise for me its a couple of lines of code so no biggie!

**Comment by kvakvs** [comment]
I'll remain quiet. It looks OK, visual touch like a contrast outline can be added later in a 2 line PR.
Do not worry :)

**Comment by kianzarrin** [comment]
Do we need to wait for @krzychu124 or can we merge this in?

**Comment by krzychu124** [comment]
Why you are so hurry 😄 Finally I got some free time so I will finish all pending tasks today, including testing this PR 😉

**Comment by aubergine10** [comment]
Is latest commits pushed?

**Comment by kianzarrin** [comment]
Yes commit [88ae11c](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635/commits/88ae11c54bc56ec2f4009e592b0d1c38ee745695) . why do you ask?

**Comment by aubergine10** [comment]
> Yes commit 88ae11c . why do you ask?

Doing some more playtesting in an hour or so and just wanted to make sure it was up to date.

**Comment by kianzarrin** [comment]
- since the sheath overlay engulfs the node marker there is no need to raycast the node marker ( a bit of speed boost and smaller code)
- with the sheath overlay flickering is not an issue so I removed the anti-flickering code.
- I render node marker after lane overlay so I get a better sheath overlay.
- tested it: no flickering no problems :)

Screenshot (415)

**Comment by aubergine10** [comment]
Looking good!! Will test after I finish "yet another page of documentation" - so about 15 mins...

**Comment by aubergine10** [comment]
YESSSSSS!!!!!! <3

image

I bet you can guess what I'll wish for as future work once this gets merged :) Back-tracing the incoming lane to show what can reach it (up to next junction), and forward tracing the outgoing lane to see where it can lead to (up to next junction). But that's future stuff.

I'm so loving how this is working now!!

**Comment by aubergine10** [comment]
While I have an incoming lane selected, could we retain the sheaths on the outgoing lanes it links to?

image

That way I would clearly see the outgoing routes connected to the currently selected incoming lane. So in the image above, there would be 2 white sheaths on the two outgoing lanes linked to that currently selected incoming lane.

**Comment by aubergine10** [comment]
Here is a better illustration of why I'd like outgoing sheaths shown for currently selected incoming lane.

image

I have red icoming lane connected to 2 outgoing lanes, but only one is clearly visible. If both connected outgoing lanes had sheath, it would be very obvious which lanes red incoming lane connects to.

If I deselect the red incoming lane, the outgoing lanes it connects to would remove their sheaths. Hope that makes sense. :)

**Comment by aubergine10** [comment]
Downside to showing all outgoing sheaths for current incoming lane is when I want to remove an existing connection.

Maybe if connected-but-not-hovered outgoing lane was given same colour sheath as the incoming lane. If I hover an outgoing lane, it's sheath would turn white, whether connected or not?

Maybe this is something to ponder in a future PR, I'll leave it up to you to decide.

**Comment by aubergine10** [comment]
> Downside to showing all outgoing sheaths for current incoming lane is when I want to remove an existing connection.
> 
> Maybe if connected-but-not-hovered outgoing lane was given same colour sheath as the incoming lane. If I hover an outgoing lane, it's sheath would turn white, whether connected or not?

Or....

When hovering non-connected outgoing lane, it gets white sheath. When connected to outgoing lane, the sheath remains white (while assocaited incoming lane still selected). But if I then hover an already-connected outgoing lane (for currently selected incoming lane), the outgoing sheath takes on colour of the incoming lane indicating that if I click that outgoing lane the wire is going to get reeled back in to the incoming lane (ie. disconnected from that outgoing lane).

Same could be done with the "wire": Dragged wire (following mouse) = white = "stick me in a white hole / sheath". Connected wire = white = distinguishes the currently active wires (for selected incoming lane) from any others on the junction/node (as they are all coloured). But if I then hover over alread-connected (to currently selected incoming lane) outgoing lane, the wire turns same colour as incoming lane = "that's where it's going if you click now - back to the coloured lane" = connection will break.

Hope that makes even a remote amount of sense lol

**Comment by aubergine10** [comment]
Minor buglet (can be left to future PR) - station track source sheaths are white: 
image

It's bit of weird scenario because station tracks are both incoming and outgoing lanes.

**Comment by kianzarrin** [comment]
> When hovering non-connected outgoing lane, it gets white sheath. When connected to currentluy-selected incoming lane, the outgoing sheath remains white (while assocaited incoming lane still selected). But if I then hover an already-connected outgoing lane (for currently selected incoming lane), the outgoing sheath takes on colour of the incoming lane indicating that if I click that outgoing lane the wire is going to get reeled back in to the incoming lane (ie. disconnected from that outgoing lane).

incoming is ambiguous. is it incoming lane from the junction or toward the junction? depends on where the observer is. is it source? is it target?

**Comment by kianzarrin** [comment]
> 
> 
> Minor buglet (can be left to future PR) - station track source sheaths are white:
> image
> EDIT: That's train station using RWY (Railway) rail tracks and station tracks.
> 
> It's bit of weird scenario because station tracks are both incoming and outgoing lanes.

Well spotted! I changed half a line of code. should be fixed now:
Screenshot (417)
Screenshot (419)
Screenshot (418)

**Comment by kianzarrin** [comment]
@aubergine10 

> Here is a better illustration of why I'd like outgoing sheaths shown for currently selected incoming lane.
> image
> I have red incoming lane connected to 2 outgoing lanes, but only one is clearly visible. If both connected outgoing lanes had sheath, it would be very obvious which lanes red incoming lane connects to.
> If I deselect the red incoming lane, the outgoing lanes it connects to would remove their sheaths. Hope that makes sense. :)

Oops almost didn't see this one. github confuses me. sometimes I don't see messages.
Fixed it! I draw the lane curves to the selected marker last so that they are on top. also they are emboldened with different (grey)  border color.

Screenshot (421)

Screenshot (420)

We can discuss all those complex Lane sheath stuff in another PR if that is OK.

**Comment by aubergine10** [comment]
I'll create an issue with all the extra ideas for future PR.

**Comment by aubergine10** [comment]
@krzychu124 any feedback on code, etc? this PR is read for merge pending your review

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 6]
```diff
@@ -54,19 +57,149 @@ private class NodeLaneMarker {
...
+                float size = 1f / n;
+                for (int i = 0; i < n; i++) {
+                    Bezier3 bezier = bezier0.Cut(i * size, (i + 1) * size);
+                    Bounds bounds = bezier.GetBounds();
+                    bounds.Expand(1f);
+                    this.bounds[i] = bounds;
+                }
+            }
+
+            internal void RenderOverlay(RenderManager.CameraInfo cameraInfo, Color color, bool enlarge=false) {
```
Comment (kvakvs) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: `///` What does, parameter docs

File: TLM/TLM/UI/TrafficManagerTool.cs [hunk 7]
```diff
@@ -1,24 +1,24 @@
 namespace TrafficManager.UI {
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using API.Manager;
+    using API.Traffic.Data;
```
Comment (kvakvs) [inline TLM/TLM/UI/TrafficManagerTool.cs]: PLease return all imports to full qualified imports.
If you suggest an option to the auto-fixer, can you also commit that change?

File: TLM/TLM/UI/TrafficManagerTool.cs [hunk 8]
```diff
@@ -968,7 +983,25 @@ public class TrafficManagerTool
             return minSegId;
         }
 
-         /// <summary>
+        private static float prev_H = 0f;
+        private static float prev_H_Fixed;
+        internal static float GetAccurateHitHeight() {
```
Comment (kvakvs) [inline TLM/TLM/UI/TrafficManagerTool.cs]: Add `///` what does

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs (hunk 6 above)
Comment (kianzarrin) [reply TLM/TLM/UI/SubTools/LaneConnectorTool.cs]: the parameters are self explanatory.

File: TLM/TLM/UI/TrafficManagerTool.cs (hunk 7 above)
Comment (kianzarrin) [reply TLM/TLM/UI/TrafficManagerTool.cs]: no these are just merge problems.

**Comment by kvakvs** [comment]
Nice

**Comment by aubergine10** [comment]
@krzychu124 this is ready to merge - any comments?

**Comment by aubergine10** [comment]
When I click on this blue lane:

image

The exiting line connecting to the outgoing lane goes mental:

image

Green lane is same:

image

And so on...

Station tracks are fixed though :)

image

Looks like the station track fix broke all the other types of lane connector...

**Comment by kianzarrin** [comment]
> @aubergine10 Looks like the station track fix broke all the other types of lane connector...
It can't be! I tested it and it worked ... look: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/635#issuecomment-583726108

Ill test it out again.

**Comment by krzychu124** [comment]
Testing...

**Pull request merged**